### PR TITLE
Add a new AssertionMatcher that reuses AssertJ assertions

### LINKED ISSUE #926: AssertionMatcher proposal
#### Summary

I have a proposal to add a generic Hamcrest Matcher implementation that would fit assertj library:

```java
    public abstract class AssertionMatcher<T> extends BaseMatcher<T> {
        private AssertionError lastError;

        @Override
        public boolean matches(Object argument) {
            T actual = (T) argument;
            try {
                assertion(actual);
                return true;
            } catch (AssertionError e) {
                lastError = e;
                return false;
            }
        }

        public abstract void assertion(T actual) throws AssertionError;

        @Override
        public void describeTo(Description description) {
            //Describe lastError and its stacktrace
        }
    }
```

Advantages as I see it:

- don't create complex boolean expressions inside matches method
- override only assertion method
- inside assertion method use assertj default, custom or any other assertions that throw AssertionError
- failure description is provided with detailed description get from assertion and with a stacktrace that points exactly to line where assertion has failed.

#### Example

Basic usage is simple, for example with Mockito library:

```java
    verify(customerRepository).save(argThat(new AssertionMatcher<Customer>() {
            @Override
            public void assertion(Customer actual) throws AssertionError {
                assertThat(actual).hasName("Tom")
                    .hasAge(30);
            }
    }));
```

#### Java 8 specific ?

* NO  : create Pull Request from the `2.x` branch

I this this proposal has generic use case and fits well with assertj library. Please review it and if you agree I would provide a PR.

### PR DESCRIPTION
**Pull request by SpOOnman**
#### Check List:
* Fixes #926 - AssertionMatcher implementation proposal
* Unit tests : YES
* Javadoc with a code example (API only) : YES

I'm not quite happy with a test AssertionMatcher_matches_Test#should_fill_description_when_assertion_fails. I've explained in code comments. I feel like I can't alter static Failures instance in this test because it would alter other tests.

Please review my contribution. Thanks!

### CODE REVIEW

File: src/main/java/org/assertj/core/matcher/AssertionMatcher.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * Generic Hamcrest {@link Matcher} that reuses AssertJ assertions.
+ *
+ * Overriding classes should only override {@link AssertionMatcher#assertion(Object)}
+ * method. {@link Matcher#matches(Object)} and {@link Matcher#describeTo(Description)}
+ * are provided. If matcher fails it will provide description that contains
+ * stacktrace of a failed assertion.
+ *
+ * <p>
+ * Example with Mockito use case:
+ * <pre><code class='java'>verify(customerRepository).save(argThat(new AssertionMatcher&lt;Customer&gt;() {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/matcher/AssertionMatcher.java]: add a space before `verify` to make javadoc indentation perfect

File: src/main/java/org/assertj/core/matcher/AssertionMatcher.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void describeTo(Description description) {
+    if (lastError != null) {
+      description.appendText("AssertionError with message: ");
+      description.appendText(lastError.getMessage());
+      description.appendText("\n\nStacktrace was: ");
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/matcher/AssertionMatcher.java]: Prefer `String.format("%n%nStacktrace was: ")` to be OS agnostic.

**Comment by joel-costigliola** [comment]
You can alter the stack trace filtering boolean as long as you set it back to its original value (using `Failures` methods). 
AssertJ tests are not run in parallel, it should not cause any failures.

File: src/main/java/org/assertj/core/matcher/AssertionMatcher.java (hunk 1 above)
Comment (SpOOnman) [inline src/main/java/org/assertj/core/matcher/AssertionMatcher.java]: Done.

File: src/main/java/org/assertj/core/matcher/AssertionMatcher.java (hunk 2 above)
Comment (SpOOnman) [inline src/main/java/org/assertj/core/matcher/AssertionMatcher.java]: Done.

**Comment by SpOOnman** [comment]
I use setUp and tearDown methods to ensure that removeAssertJRelatedElementsFromStackTrace is set back to its original value.

**Comment by joel-costigliola** [comment]
AssertionMatcher_matches_Test.should_fill_description_when_assertion_fails failed in travis : https://travis-ci.org/joel-costigliola/assertj-core/builds/205524692

**Comment by SpOOnman** [comment]
I don't know why but line numbers were different. I've removed line numbers. Still, I don't know why there were different. I've amended my commit and now it passes.

**Comment by SpOOnman** [comment]
Hi, any news on this? Is there anything I can improve?

**Comment by joel-costigliola** [comment]
Will do a final review in the we

**Comment by joel-costigliola** [comment]
Integrated thanks  @SpOOnman !

**Comment by SpOOnman** [comment]
Thank you!

**Pull request closed without merging**